src/voice-broadcast/models/VoiceBroadcastRecording.ts

```diff
@@ -248,56 +320,102 @@ export class VoiceBroadcastRecording
     };
 
     private onChunkRecorded = async (chunk: ChunkRecordedPayload): Promise<void> => {
-        const { url, file } = await this.uploadFile(chunk);
-        await this.sendVoiceMessage(chunk, url, file);
+        const uploadAndSendFn = async () => {
+            const { url, file } = await this.uploadFile(chunk);
+            await this.sendVoiceMessage(chunk, url, file);
+        };
+
+        try {
+            await uploadAndSendFn();
+        } catch {
+            // Uploading audio failed.
+            this.toRetry.push(uploadAndSendFn);
+            await this.onConnectionError();
+        }
     };
 
-    private uploadFile(chunk: ChunkRecordedPayload): ReturnType<typeof uploadFile> {
+    /**
+     * This function is called on connection errors.
+     * It sets the connection error state and stops the recorder.
+     */
+    private async onConnectionError(): Promise<void> {
+        await this.stopRecorder();
+        this.setState("connection_error");
+    }
+
+    private async uploadFile(chunk: ChunkRecordedPayload): ReturnType<typeof uploadFile> {
         return uploadFile(
             this.client,
-            this.infoEvent.getRoomId(),
+            this.roomId,
             new Blob([chunk.buffer], {
                 type: this.getRecorder().contentType,
             }),
         );
     }
 
-    private async sendVoiceMessage(chunk: ChunkRecordedPayload, url: string, file: IEncryptedFile): Promise<void> {
-        const content = createVoiceMessageContent(
-            url,
-            this.getRecorder().contentType,
-            Math.round(chunk.length * 1000),
-            chunk.buffer.length,
-            file,
-        );
-        content["m.relates_to"] = {
-            rel_type: RelationType.Reference,
-            event_id: this.infoEvent.getId(),
-        };
-        content["io.element.voice_broadcast_chunk"] = {
+    private async sendVoiceMessage(chunk: ChunkRecordedPayload, url?: string, file?: IEncryptedFile): Promise<void> {
+        const sendMessageFn = async () => {
+            const content = createVoiceMessageContent(
+                url,
+                this.getRecorder().contentType,
+                Math.round(chunk.length * 1000),
+                chunk.buffer.length,
+                file,
+            );
+            content["m.relates_to"] = {
+                rel_type: RelationType.Reference,
+                event_id: this.infoEventId,
+            };
             /** Increment the last sequence number and use it for this message. Also see {@link sequence}. */
-            sequence: ++this.sequence,
+            content["io.element.voice_broadcast_chunk"] = {
+                sequence: ++this.sequence,
+            };
+
+            await this.client.sendMessage(this.roomId, content);
         };
 
-        await this.client.sendMessage(this.infoEvent.getRoomId(), content);
+        try {
```

Comment:
Could this try catch be abstracted into a wrapping function? something like
```
private async callWithRetry(function) => {
    try {
        await function();
    catch {
       this.toRetry.push(function);
       this.onConnectionError
    }
}
// then something like:
await this.callWithRetry(() => this.sendVoiceMessage(props));

```

Reply:
Factored out `callWithRetry` :heavy_check_mark: